Steps towards a StorageEngine API

chrisvest opened the pull request:
We want to introduce a StorageEngine API, that decouples the mechanics of turning graph IO into byte IO, from the mechanics of a full-blown database. In essence, we would like an API for non-transactionally reading a graph store, and for turning transaction state into a stream of store change commands, and for applying said stream of change commands to the store.

This way, storage gets decoupled from the rest of the database, allowing us to have more than one storage implementation. The refactorings in this PR moves us in this direction.

chrisvest committed:
Push creation of StoreFactory down into NeoStoreDataSource

chrisvest committed:
Move the StoreFactory construction closer to the StoreLayerModule

Because it is only the StoreLayerModule that really needs it.
This is another step towards better encapsulation.

chrisvest committed:
Remove a few uses of Neo4j-specific Provider & Supplier interfaces

Replaced with the `java.util.function.Supplier` interface.

jakewins commented on community/function/src/main/java/org/neo4j/function/Supplier.java:
```diff
@@ -24,6 +24,7 @@
  *
  * @param <T> the type of results supplied by this supplier
  */
+@Deprecated // Use java.util.function.Supplier instead
```
If we put these deprecations in 2.3.x, we can remove this altogether in 3.0..

chrisvest replied:
Good idea. I'll pull them out and send them to 2.3.

thobe replied:
@jakewins I'd argue that only to be true for `2.3.0`, at least if it is public API.
If it is private, we can do whatever we want with it, whenever we want to.

chrisvest committed:
Reduce NeoStoreDataSource coupling to NeoStoresModule

Including removing it as an explicit field.
The responsibility of managing the NeoStores is gradually moving into the StoreLayerModule, which will get a better name in due time.

chrisvest committed:
Merge IndexingModule into StoreLayerModule in NeoStoreDataSource

Once all storage related things have been collected into a single interface, we can start encapsulating by pushing responsibility down.

chrisvest committed:
Fix DumpProcessInformationTest

It was occasionally using the wrong version of Java, when you had more than one version installed.
Specifically, if the version of Java that the classes had been compiled for, was different than the version that was responding to the `java` command by default.

chrisvest committed:
Remove the NeoStoreModule from NeoStoreDataSource and abstract UpdatableSchemaState out of the Index stuff into a callback

This way, the indexes don't need to know anything about caches, or how their invalidation is necessary when an index comes online.

chrisvest committed:
Move SchemaCache and CacheAccessBackdoor into the StoreLayerModule

chrisvest committed:
Move the procedure cache into the StoreLayerModule

It had nothing to do with the other cache stuff in the CacheModule.
It should probably be encapsulated as a store, and the fact that it's currently just a "cache" should be made transparent.
But that's for later.

chrisvest committed:
Begin the process of extracting the StoreLayerModule into a StorageEngine

Currently only the parts necessary to build a StoreReadLayer have been pulled into the StorageEngine abstraction.
Everything around creating and applying commands is still missing, as well as a lot of work in encapsulating the internals of it.

chrisvest committed:
Pull dependencies into StorageEngine and make dependants go through that API for now

Later, the hope is that we can push dependent code down into the StorageEngine to encapsulate.

chrisvest committed:
Congregate more dependencies in the StorageEngine and introduce a CommandStream abstraction to decouple store application from transaction representation

chrisvest committed:
Move the construction of IndexConfigStore into NeoStoreDataSource

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/storageengine/StorageEngine.java:
```diff
@@ -0,0 +1,84 @@
+    TransactionIdStore transactionIdStore();
+    LogVersionRepository logVersionRepository();
+
+    @Deprecated
```
Why all those methods are deprecated?

tinwelint replied:
Because they are temporary in this overhaul that is the introduction of the storage engine API

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/CommandStream.java:
```diff
@@ -0,0 +1,39 @@
+ * A stream of commands from one or more transactions, that can be serialised to a transaction log or applied to a
+ * store.
+ */
+public interface CommandStream
```
Can we please avoid using confusing names for things that we just introduce. Here we introduce a stream which is not a stream at all, i think we already have a wide range of confusing stuff in a code base.

chrisvest replied:
It's too early to decide if we want it to implement the Java Stream interface or not. Stream is used in a more conceptual way, here. We'll revise the name when the transformation of transaction state into commands, and command application, gets pushed into the storage engine.

MishaDemianenko replied:
in a what ever way stream is used here it should not be confusing :), since it will be revisited perfectly fine. Keep silence for now :no_mouth:

MishaDemianenko commented:
@chrisvest do we have anywhere description of what is a final goal of PRs  from this chain?

chrisvest commented:
@MishaDemianenko Yes, the description is on the StorageEngine card on our wall.

MishaDemianenko commented:
@chrisvest @tinwelint thx guys that gives some insight
